cyto_dl/models/jepa/iwm.py

```diff
@@ -0,0 +1,121 @@
+        max_epochs : int, optional
+            The maximum number of training epochs (default is 100).
+        predict_target : bool, optional
+            Whether to predict the target embeddings instead of just extracting embeddings of source image (default is False).
+        **base_kwargs : dict
+            Additional arguments passed to the BaseModel.
+        """
+        super().__init__(
+            encoder=encoder,
+            predictor=predictor,
+            x_key=source_key,
+            save_dir=save_dir,
+            momentum=momentum,
+            max_epochs=max_epochs,
+            **base_kwargs,
+        )
+
+    def model_step(self, stage, batch, batch_idx):
+        self.update_teacher()
+        source = batch[self.hparams.source_key]
+        target = batch[self.hparams.target_key]
+
+        target_masks = self.get_mask(batch, "target_mask")
+        context_masks = self.get_mask(batch, "context_mask")
+        target_embeddings = self.get_target_embeddings(target, target_masks)
+        context_embeddings = self.get_context_embeddings(source, context_masks)
+        predictions = self.predictor(
+            context_embeddings, target_masks, batch[self.hparams.target_domain_key]
+        )
+
+        loss = self.loss(predictions, target_embeddings)
+        return loss, None, None
+
+    def get_predict_masks(self, batch_size, device):
+        mask = torch.ones(self.hparams.encoder["num_patches"], dtype=bool, device=device)
+        mask = rearrange(mask, "z y x -> (z y x)")
+        mask = torch.argwhere(mask).squeeze()
+
+        return repeat(mask, "t -> t b", b=batch_size)
+
+    def extract_embeddings(self, tensor):
+        # mean across patches, no cls token to remove
+        return tensor.mean(axis=1).detach().cpu().numpy()
+
+    def predict_step(self, batch, batch_idx):
+        source = batch[self.hparams.source_key]
+        source = self.remove_first_dim(source)
+
+        embeddings = self.encoder(source)
+        if self.hparams.predict_target:
+            # use predictor to predict each patch
+            target_masks = self.get_predict_masks(source.shape[0], device=source.device)
+            # predict target embeddings from source embeddings
+            embeddings = self.predictor(
+                embeddings, target_masks, batch[self.hparams.target_domain_key]
+            )
+        return self.extract_embeddings(embeddings), source.meta
+
+    def test_step(self, batch, batch_idx):
+        source = batch[self.hparams.source_key]
+        source = self.remove_first_dim(source)
+        target = batch[self.hparams.target_key]
+        target = self.remove_first_dim(target)
+
+        # use predictor to predict each patch
+        target_masks = self.get_predict_masks(source.shape[0], device=source.device)
+
+        # mean across patches, no cls token to remove
+        source_embeddings = self.encoder(source)
```

Comment:
is the encoder here a vit? Does it handle the patchifying etc? Is the source here the small mask?

Reply:
yes, the vit encoder does patchifying. source is an image. This is taking all the patches from the source image and predicting the embeddings of all the patches from the target image.